src/main/java/io/vertx/core/net/impl/pool/CombinerExecutor.java

```diff
@@ -72,23 +82,42 @@ public void submit(Action<S> action) {
       }
     } while (!q.isEmpty() && s.compareAndSet(0, 1));
     if (head != null) {
-      InProgressTail inProgress = current.get();
+      InProgressTail<S> inProgress = (InProgressTail<S>) current.get();
       if (inProgress == null) {
-        inProgress = new InProgressTail();
+        inProgress = new InProgressTail<>(this, tail);
         current.set(inProgress);
-        inProgress.task = tail;
         try {
           // from now one cannot trust tail anymore
           head.runNextTasks();
+          assert inProgress.others == null || inProgress.others.isEmpty();
         } finally {
           current.remove();
         }
       } else {
-        assert inProgress.task != null;
-        Task oldNextTail = inProgress.task.replaceNext(head);
-        assert oldNextTail == null;
-        inProgress.task = tail;
-
+        if (inProgress.combiner == this) {
+          Task oldNextTail = inProgress.task.replaceNext(head);
+          assert oldNextTail == null;
+          inProgress.task = tail;
+        } else {
+          Map<CombinerExecutor<S>, Task> map = inProgress.others;
+          if (map == null) {
+            map = inProgress.others = new HashMap<>();
```

Comment:
make it as small as possible, like capacity 2 or 1, depends on the default load factor